toolchain/check/context.cpp

```diff
@@ -799,30 +801,79 @@ auto Context::SetBlockArgResultBeforeConstantUse(SemIR::InstId select_id,
-                 "No current function, but node_id not provided");
-    TODO(node_id,
-         "Control flow expressions are currently only supported inside "
-         "functions.");
+auto Context::AddToRegion(SemIR::InstBlockId block_id) -> void {
+  if (region_stack_.empty() || block_id == SemIR::InstBlockId::Unreachable) {
     return;
   }
 
-  if (!inst_block_stack().is_current_block_reachable()) {
-    // Don't include unreachable blocks in the function.
-    return;
+  region_stack_.back().block_ids.push_back(block_id);
+}
+
+auto Context::BeginSubpattern() -> void {
+  auto entry_block_id = sem_ir().inst_blocks().AddDefaultValue();
+  inst_block_stack().Push(entry_block_id);
+  PushRegion(entry_block_id);
+}
+
+auto Context::EndSubpatternAsExpression(SemIR::InstId result_id)
+    -> SemIR::RegionId {
+  // TODO: Is it possible to validate that this region is genuinely
+  // single-entry, single-exit?
+  SemIR::Region region = PopRegion(result_id);
+  CARBON_CHECK(region.block_ids.back() == inst_block_stack().PeekOrAdd());
+  if (region.block_ids.size() > 1) {
+    // End the exit block with a branch to a successor block, whose contents
+    // will be determined later.
+    AddInst(SemIR::LocIdAndInst::NoLoc<SemIR::Branch>(
+        {.target_id = inst_blocks().AddDefaultValue()}));
+  } else {
+    // This single-block region will be inserted as a SpliceBlock, so we don't
+    // need control flow out of it.
   }
+  inst_block_stack().Pop();
+  return sem_ir().regions().Add(region);
+}
+
+auto Context::EndSubpatternAsEmpty() -> void {
+  auto region = region_stack_.pop_back_val();
+  CARBON_CHECK(region.block_ids.front() == inst_block_stack().PeekOrAdd());
+  CARBON_CHECK(inst_block_stack().PeekCurrentBlockContents().empty());
+  inst_block_stack().PopAndDiscard();
+}
 
-  auto function_id =
-      insts()
-          .GetAs<SemIR::FunctionDecl>(return_scope_stack().back().decl_id)
-          .function_id;
-  functions()
-      .Get(function_id)
-      .body_block_ids.push_back(inst_block_stack().PeekOrAdd());
+auto Context::InsertHere(SemIR::RegionId region_id) -> SemIR::InstId {
+  auto region = sem_ir_->regions().Get(region_id);
+  auto loc_id = insts().GetLocId(region.result_id);
+  auto exit_block = inst_blocks().Get(region.block_ids.back());
+  if (region.block_ids.size() == 1) {
+    // TODO: Is it possible to avoid leaving an "orphan" the block in IR in the
+    // first two cases?
+    if (exit_block.size() == 0) {
+      return region.result_id;
+    } else if (exit_block.size() == 1) {
+      inst_block_stack_.AddInstId(exit_block.front());
+      return region.result_id;
+    } else {
+      return AddInst<SemIR::SpliceBlock>(
+          loc_id, {.type_id = insts().Get(region.result_id).type_id(),
+                   .block_id = region.block_ids.front(),
+                   .result_id = region.result_id});
+    }
+  } else {
+    if (region_stack_.empty()) {
+      return SemIR::ErrorInst::SingletonInstId;
+    }
```

Comment:
> > The problem with diagnosing it here is we don't have a good `LocId` to diagnose on: we want to say something like "Semantics TODO: control flow expression outside a function body" and then point to the control flow expression, but in code like `x: if true then i32 else i64`, the `LocId` for the binding pattern's type expression is getting lost somewhere in the very long chain of custody that should bring it here, whereas that's not an issue if we stick with the status quo of diagnosing in the handler for the `IfExprElseId` parse node.

Update: it looks like the location only gets lost when the type expression involves a short-circuiting boolean operator; it works fine with the conditional operator.

> How would you feel about calling `Context::TODO` here before returning `ErrorInst`, and adding a matching call to `AddToRegion` for consistency?

I don't feel great about it. To illustrate why, I've made that change, but I've temporarily added prefixes to the error messages to indicate where they're coming from (I've also added a couple test cases to show corners that weren't already covered). As you can see, it leads to a _lot_ of redundant noise in the diagnostics.

Note that I have a followup PR to switch `let` and `var` declarations to use the new pattern-matching framework, which will cause `InsertHere` to start diagnosing on `and_` and `or_` in `fail_and_or_not_in_function.carbon` (but without a valid location, as noted) and on the `y` in `fail_not_in_function.carbon`, but that still doesn't give us a path to having full diagnostic coverage without excessive repetition (other than what I had to begin with):
- `InsertHere` doesn't diagnose on control flow that isn't semantically part of a pattern (e.g. `let x: i32 = if true then 1 else 0;`).
- `AddToRegion` doesn't diagnose on control flow that's syntactically inside a binding pattern (e.g. `var y: if true then i32 else f64;`).
- As a result, neither of them diagnoses on control flow in the type of a class field declaration (or an interface associated constant declaration, I expect).
- `AddToRegion` generally diagnoses multiple times if it diagnoses at all, because it diagnoses once per block (including blocks that we add as an implementation detail).
- Every case also gets exactly one diagnostic from the `IfExprElseId` handler or from `HandleShortCircuitOperator`.

Is there some way to query that a diagnostic has already been emitted? If so, I could `CHECK` on that here, to ensure that compilation won't succeed or another, without adding a redundant diagnostic.

Or maybe all the diagnostic noise is OK, since these are TODOs rather than diagnostics we expect to keep long-term?